README.md
OpenSTV: Introduction & Historical Notes
========================================

This is OpenSTV 1.7.  OpenSTV is open source voting software that implements the single transferable vote (STV) election method.

OpenSTV was released under the GNU General Public License version 2 (or later) until around mid-2011.  At that point its main author, Jeff O'Neill, took future versions proprietary.  As of this writing, the latest version along the proprietary line is 2.1.0 and it is available for purchase from http://www.openstv.org/.  It looks like O'Neill also offers an online web app at http://www.opavote.org/, in case you don't want to set up OpenSTV 2.x yourself.

The last open source version of OpenSTV he released was 1.7, which is available here.  If this were ever to become the root of a new open source development line, good manners would dictate that we change the name to something other than "OpenSTV" (unless O'Neill prefers it stay the same).  However, that's not an issue at this point, as what's here is exactly what was released as "OpenSTV 1.7" back in 2011.

For more information about OpenSTV's history, and the temporary disappearance of the 1.7 source code from the Internet, see:

  * http://groups.google.com/group/openstv-announce/browse_frm/month/2011-06?pli=1
  * http://meta.stackoverflow.com/questions/125579/downloading-openstv-doesnt-work
  * http://meta.stackoverflow.com/questions/78203/is-it-legal-for-stack-overflow-to-redistribute-openstv-binaries

Below is the original README.txt from Jeff O'Neill (just reformatted to Markdown for consistency with the rest of this file):

Original README.txt from OpenSTV 1.7
====================================

`Revision $Id: README.txt 578 2009-08-26 01:13:50Z jeff.oneill $`

Overview
--------

OpenSTV is a program for implementing the single transferable vote (STV).  STV is used for electing a group of people (e.g. council, committee, legislature) and it provides for proportional representation of the electorate.  The idea behind proportional representation is that the demographics of the elected group should, at least roughly, match the demographics of the electorate.  The beauty of STV is that there are no reserved seats and the proportional representation arises naturally.  For more information see

  * http://www.fairvote.org/
  * http://www.electoral-reform.org.uk/

In an STV election, each voter simply ranks the candidates in order of preference.  The rules for counting the votes with STV are more complicated than winner-take-all elections.  The votes can be counted by hand, but it is useful to have a computer program to speed up the process.  There are several variations of STV, but most users should use the default options.

Installation
------------

See http://www.OpenSTV.org/ for more information.

Jeff O'Neill
jeff.oneill at openstv.org


CHANGELOG.txt
$Id: CHANGELOG.txt 822 2010-11-21 05:25:43Z jeff.oneill $

OpenSTV 1.7 
- Added SanFranciscoRCV method 
- Added support for cleaning of ballots as done in San Francisco
(before was only Cambridge cleaning)
- Relaxed minimum requirements for number of candidates and ballots

OpenSTV 1.6
- Provide YAML report for integration with DemoChoice
- Added support for equal rankings and skipped rankings on ballots
- Added support for ballot IDs
- Restructuring of data structures for ballots
- Reports are now done as plugins
- Improvements to report generation including writing to file instead of memory
- Counting methods are much faster with a large number of candidates
- Check folder Plugins for external plugins
- Reorganized package structure so that all imports use absolute paths
- OpenSTV uses threads for counting and loading ballots so user doesn't see 
  a frozen screen
- Allow manual tie-breaking in OpenSTV

OpenSTV 1.5
- Improved format of election results
- No limit on the number of candidates (previously was limited to 255
candidates)
- Election methods are now done as plugins so users can add their own
methods (although only for advanced users at this point).
- Cambridge STV can save the winner's ballots in files for determining
replacements after a vacancy
- Added statutes of some rules to the documentation
- Added draft of New Zealand Meek (work in progress)
- Added new method called QPQ
- Comments now allowed in ballot files
- Changes under the hood that make OpenSTV easier to maintain.

OpenSTV 1.4
- Fixed bug that broke Election Options dialog on mac
- Fixed bug when navigating from empty ballot
- Fixed bug in remembering working directory on mac

OpenSTV 1.3
- Rewrote OpenSTV user interface:
(1) User can only run an election from a saved ballot file
(previously, you could use a ballot file in memory).
(2) User can edit/create ballots in a separate Ballot File Editor
window and save them to a file.
- Can save election results in HTML.
- More explanatory output for Condorcet elections.
- Added subclasses for different ballot types to make it easier to
support new types.
- Changes under the hood that make OpenSTV easier to maintain.

OpenSTV 1.2
- Changed output to be more descriptive and easier to read.
- Changed implementation of SuppVote to use all rankings.
- Changed implementation of IRV to stop sooner (won't change outcome).
- Small change in reporting of the threshold for ERS97 (won't change outcome).
- Print substages for ERS97.
- Now possible to have unlimited precision (but max is set to 20 for gui).
- Meek/Warren implemented in fixed point.

OpenSTV 1.1
- Added Scottish STV rules.
- Added GPCA2000 rules.
- Added approval voting.
- Bug fixes in N. Ireland STV.
- Simplified IRV.  Now always eliminates candidates one by one and
stops when candidate has a majority (instead of stopping when 2
candidates remain).
- Show help files within OpenSTV instead of external browser because
the external browser was not working with some platforms.
- BCSTV, FTSTV, and Borda now implemented in fixed point instead of
floating point.  Coombs, Meek, and Warren still in floating point.
- Major code rewrite in STV.py.

OpenSTV 1.0
- Changed name to OpenSTV!
- Test code now in a separate file and added more tests.
- Although the tree-packed ballots greatly increased speed, they also
greatly increased memory consumption.  The tree-data structure is now
created dynamically as necessary rather than built completely.  This
retains (and sometimes improves) the speed, and memory consumption is
negligable.  Condorcet no longer uses trees.
- Ballot data now based on bytes instead of ints which uses 1/4th less
memory.  The number of candidates is now limited to 255.
- Modifications to allow the use of psyco to speed up the code.  Psyco
on average speeds up the code by a factor of 2-4 times.
- Now properly raises and catches warnings and exceptions.
- Fixed bug in Condorcet completion methods.
- Changed code structure so that each method is a class rather than a 
function.  This allows for cleaner code and better code reuse.
- Fixed bugs with withdrawn candidates.  Withdrawn candidates are now
excised from a copy of the ballots before the start of the election.  
Since it is done with a copy, it can later be undone.
- "raw" ballot format renamed as "text" format because raw also means
unpacked ballots.

pSTV 0.9
Implemented the supplemental vote and N. Ireland STV.  Started Dail
STV.  Can shuffle the ballots (interesting for Cambridge STV and
random transfer STV).  Simplified the rounding of floats to make the
code cleaner.  Implemented tree-packed ballots.  Used tree-packed
ballots with Meek, Warren, and Condorcet, greatly speeding up the
implementations of these methods.  Faster implementation of Bucklin
and Coombs.  For the curious, here are the timing statistics, in
seconds, as they have changed over time on a test of data from eight
real elections:
     SNTV:         0.6
     IRV:          7.1 ->  1.2
     BCSTV:        4.6
     Coombs:      68.6 -> 40.4 -> 24.8
     Cambridge:    3.7 ->  2.0
     ERS97:        6.4 ->  5.5
     Meek:        48.3 -> 38.7 ->  5.9
     Warren:      38.0 -> 30.7 ->  3.0
     Bucklin:      9.4 ->  3.6
     Borda:        2.3
     Condorcet:  106.0 -> 13.3 ->  7.7


pSTV 0.8
Added tabbed windows for comparing multiple election results.  The
font size can be changed for each tab separately.  The display
precision is now chosen in the Election menu and can only be changed
by rerunning the election.  The ERS97 implementation is finally in
fixed point and the desired precision can be specified (for future
implementation of ERS97-like rules such as Malta STV).  Can now
eliminate candidates before running an election.  Previously, when
loading a BLT file, eliminated candidates were removed from the
ballots.  Now, the eliminated candidates are noted but they remain on
the ballots.  This works better with allowing the user to selectively

setup2.py
#!/usr/bin/env python
## Copyright (C) 2003-2010  Jeffrey O'Neill
##
## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

import sys
import os
import glob
import shutil
import pkgutil
from distutils.core import setup

from openstv.version import v as OpenSTV_version

__revision__ = "$Id: setup.py 777 2010-06-06 13:04:54Z jeff.oneill $"

# Remove the build and dist folders
shutil.rmtree("build", ignore_errors=True)
shutil.rmtree("dist", ignore_errors=True)

# generic arguments for setup()
name = "OpenSTV"
description = "Implements the single transferable vote, instant runoff "\
              "voting, and several other voting systems."
author = "Jeff O'Neill"
author_email = "jeff.oneill@openstv.org"
url = "http://www.openstv.org"
license = "Gnu General Public License version 2"

packages = ["openstv",
            "openstv.MethodPlugins", 
            "openstv.LoaderPlugins",
            "openstv.ReportPlugins"]

data_files = [ ("", ["openstv/Help.html",
                     "openstv/License.html"]),
               ("Icons",
                 ["openstv/Icons/pie.ico",
                  "openstv/Icons/blt.ico",
                  "openstv/Icons/splash.png"]) ]

# Get list of plugins because they won't be automatically included
includes = []
pluginPackageNames = ["openstv.MethodPlugins", "openstv.LoaderPlugins", 
                      "openstv.ReportPlugins"]
for x in pluginPackageNames:
    package = __import__(x, fromlist="dummy")
    ppath = package.__path__
    pname = package.__name__ + "."
    for importer, modname, ispkg in pkgutil.iter_modules(ppath, pname):
        includes.append(modname)

# For *nix, data_files goes into the /usr directory, which 
# we don't want.
# pacakge_data doesn't appear to work for sdist so use MANIFEST.in

# Create MANIFEST.in
f = open("MANIFEST.in", "w")
print >> f, r"""
include CHANGELOG.txt
include README.txt
include openstv/Help.html
include openstv/License.html
include openstv/Icons/splash.png
include openstv/Icons/blt.ico
include openstv/Icons/pie.ico
include openstv/Icons/blt.icns
include openstv/Icons/pie.icns
"""
f.close()

setup(name = name,
      version = OpenSTV_version,
      description = description,
      author = author,
      author_email = author_email,
      url = url,
      license = license,
      packages = packages,
      )
